File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let begin = 0, max = 0, map = {}
    for(let end = 0; end<s.length; end++){
        if(map[s[end]] !== undefined && map[s[end]] >= begin){
            begin = map[s[end]] + 1
        }
        let len = end - begin + 1
        max = Math.max(len, max)
        map[s[end]] = end
    }
    return max
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen_chars_indices, begin, max_len = {}, 0, 0
        for end in range(len(s)):
            while s[end] in seen_chars_indices and seen_chars_indices[s[end]] >= begin:
                begin = seen_chars_indices[s[end]] + 1
            length = end - begin + 1
            max_len = max(max_len, length)
            seen_chars_indices[s[end]] = end
        return max_len